### ISSUE TITLE
Optional Int values don't get stored in Release builds.

### ISSUE THREAD

**Issue by StarWars** [realm/realm-cocoa#5622]
## Goals
I wanted to store Optional Int values using `RealmOptional<Int>()`.

## Expected Results
I expected to be able to access the `Int?` value using item.value

## Actual Results
The result for the DEBUG build works as expected. The result for the Release build is that the `.value` always returns `nil`.

## Steps to Reproduce
1. Create an object. 
2. Configure one of it's fields to be an optional Int (`RealmOptional<Int>()`). 
3. Try to read the optional value.

## Version of Realm and Tooling

Realm framework version: The latest `3.1.1`

Realm Object Server version: ?

Xcode version: 9.2

iOS/OSX version: 11.2+

Dependency manager + version: Cocoapods 1.4.0

**Comment by bdash** [comment]
We have tests that verify this functionality works in both debug and release builds. Please share a more complete code snippet that demonstrates how to reproduce this problem.

**Comment by StarWars** [comment]
I am not sure what's the best approach here, since the Realm Browser shows that the optional value is in the database, using `po object.optional` in `lldb` presents the object with the value in it, but the `po object.optional.value` returns nil when in Release build. The Debug build works properly.

Here is the configuration of my realm db:

    ```
private let currentSchemaVersion: UInt64 = 1
private(set) lazy var configuration: Realm.Configuration = {
        return Realm.Configuration(schemaVersion: currentSchemaVersion, migrationBlock: nil, deleteRealmIfMigrationNeeded: true)
    }()
```

**Comment by bdash** [comment]
Have you tested what value is returned using code rather than expressions in LLDB? Can you share the declaration of the model class in question?

**Comment by StarWars** [comment]
Of course, I have tested the app first, which led me to the issue in the first place. Here is the part of the model:

```
import Foundation
import RealmSwift

@objcMembers
class RNotWorking: PersistedObject {

    dynamic var id = RealmOptional<Int>()

    dynamic var name: String?

    dynamic var someOtherObjects = List<RSomeOtherObj>()

    dynamic var someObjects = List<RSomeObj>()

    dynamic var defaultZoomLevel = RealmOptional<Int>()

    dynamic var someOtherObj: RSomeOtherObj?

    convenience init (id: Int, name: String, someOtherObjects: [RSomeOtherObj, defaultZoomLevel: Int, someOtherObj: RSomeOtherObj?, someObjects: [RSomeObj]) {
        self.init()
        self.id.value = id
        self.name = name
        self.someOtherObjects.append(objectsIn: someOtherObjects)
        self.defaultZoomLevel.value = defaultZoomLevel
        self.someOtherObj = someOtherObj
        self.someObjects.append(objectsIn: someObjects)
    }

    override static func primaryKey() -> String? {
        return "id"
    }

}
```

And here is the PersistedObject class

```
import Realm
import RealmSwift
import UIKit

class PersistedObject: Object {

    required init(realm: RLMRealm, schema: RLMObjectSchema) {
        super.init(realm: realm, schema: schema)
    }

    required init() {
        super.init()
    }

    required init(value: Any, schema: RLMSchema) {
        fatalError("init(value:schema:) has not been implemented")
    }

}
```

**Comment by bdash** [comment]
Why is `PersistedObject` overriding those initializers, including overriding one to just call `fatalError`? That's going to cause you surprising behavior, even if it may not be relevant here. It shouldn't ever be necessary to override any `init` methods of `Object` that work in terms of the Realm Objective-C types.

I dropped your class into a test app, removing the model properties that reference other types that you've not shared, and wrote some basic code using it (writing an instance of the type to the Realm, then retrieving it). I was not able to reproduce the problem you describe. Can you please share a complete code snippet that you've confirmed reproduces the problem rather than a partial snippet where we're left to guess at what else you're doing?

**Comment by StarWars** [comment]
I checked the git history, and these were the leftovers from the JSON mapper. Sorry about putting them here.
I didn't find anything relevant to the problem in the rest of the code, which is why I didn't put it here. The problem with id being nil when querying for it's value comes from the objects that were treated with the plain CRUD:

```
 func saveRNotWorking(_ configurations: [RNotWorking]) {
        try! self.realmObj?.write {
            self.realmObj?.add(configurations, update: true)
        }
}

func getRNotWorking() -> [RNotWorking] {
        if let configurations = realmObj?.objects(RNotWorking.self) {
            return Array(configurations)
        } else {
            return []
        }
    }
```
and finally: 

```
fileprivate var configurations: [RNotWorking] {
        return RealmHelpers().getRNotWorking()
}
// Here is the problem
self.configurations[indexPath.row].id?.value -> returns nil

```
The number of `RNotWorking` objects is correct so the tableview populates properly. 
The `self.configurations[indexPath.row]` was the expected object.
The value of the `id` field was stored in the DB (as seen using Realm Browser), but the `id?.value` call returned nil. 
The most surprising fact is that `id` itself was not a nil.

I fixed my implementation by replacing the `RealmOptional<Int>` with `dynamic var id = 0`, and everything worked as expected, so I don't suspect my CRUD. 
I am really sorry that I am not much of a help, I wish I could pinpoint the exact issue, but that's as much as I can figure out without the Realm sources.
